src/runtime/recipe/recipe-walker.ts

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {ConnectionConstraint} from './connection-constraint.js';
+import {Handle} from './handle.js';
+import {HandleConnection} from './handle-connection.js';
+import {Particle} from './particle.js';
+import {Recipe} from './recipe.js';
+import {Slot} from './slot.js';
+import {SlotConnection} from './slot-connection.js';
+import {SlotSpec} from '../particle-spec';
+import {Walker, WalkerTactic} from './walker.js';
+
+export class RecipeWalker extends Walker {
+
+  // Optional lifecycle events
+
+  // tslint:disable-next-line: no-any
+  onHandle?(recipe: Recipe, handle: Handle): any;
+  // tslint:disable-next-line: no-any
+  onHandleConnection?(recipe: Recipe, handleConnection: HandleConnection): any;
+  // tslint:disable-next-line: no-any
+  onParticle?(recipe: Recipe, particle: Particle): any;
+  // tslint:disable-next-line: no-any
+  onRecipe?(recipe: Recipe, result): any;
+  // tslint:disable-next-line: no-any
+  onPotentialSlotConnection?(recipe: Recipe, particle: Particle, slotSpec: SlotSpec): any;
+  // tslint:disable-next-line: no-any
+  onSlotConnection?(recipe: Recipe, slotConnection: SlotConnection): any;
+  // tslint:disable-next-line: no-any
+  onSlot?(recipe: Recipe, slot: Slot): any;
+  // tslint:disable-next-line: no-any
+  onObligation?(recipe: Recipe, obligation: ConnectionConstraint): any;
+
+  onResult(result) {
+    super.onResult(result);
+    const recipe: Recipe = result.result;
+    const updateList = [];
+
+    // update phase - walk through recipe and call onRecipe,
+    // onHandle, etc.
+
+    if (this.onRecipe) {
+      result = this.onRecipe(recipe, result);
```

Comment:
Do you want to override the argument or create a new variable? In below lines you're creating new variables, here you override the argument. I assume you want the former, but the naming clash is unfortunate.

I'm thinking whether we could rename `onResult`/`onResultDone` to something like `onItem`/`onItemDone`. From perspective of this method, the `result` is something that will be created by applying the update function to the item passed in. WDYT?